core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -610,24 +583,95 @@ public IndexResult index(Index index) throws IOException {
         }
     }
 
-    private IndexResult indexIntoLucene(Index index, long seqNo, long currentVersion, boolean deleted, boolean forceUpdateDocument, boolean canOptimizeAddDocument, long expectedVersion) throws IOException {
+    private IndexingPlan planIndexingAsNonPrimary(Index index) throws IOException {
+        final IndexingPlan plan;
+        if (canOptimizeAddDocument(index) && mayHaveBeenIndexedBefore(index) == false) {
+            // no need to deal with out of order delivery - we never saw this one
+            assert index.version() == 1L :
+                "can optimize on replicas but incoming version is [" + index.version() + "]";
+            plan = IndexingPlan.optimizedAppendOnly(index.seqNo());
+        } else {
+            // drop out of order operations
+            assert index.versionType().versionTypeForReplicationAndRecovery() == index.versionType() :
+                "resolving out of order delivery based on versioning but version type isn't fit for it. got ["
+                    + index.versionType() + "]";
+            // unlike the primary, replicas don't really care to about creation status of documents
+            // this allows to ignore the case where a document was found in the live version maps in
+            // a delete state and return false for the created flag in favor of code simplicity
+            final LuceneDocStatus luceneOpStatus = checkLuceneDocStatusBasedOnVersions(index);
```

Comment:
so if this returns `LuceneDocStatus.NEWER_OR_EQUAL` then the existing doc is newer or equal and not the given doc? that is very confusing, I mentioned this before I think. I'd never expect that. The opposite is intuitive?